daemon/cmd/ciliumendpoints.go

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cilium/cilium/pkg/endpoint"
+	cilium_v2a1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+)
+
+type localEndpointCache interface {
+	LookupPodName(name string) *endpoint.Endpoint
+}
+
+// This must only be run after K8s Pod and CES/CEP caches are synced and local endpoint restoration is complete.
+func (d *Daemon) cleanStaleCEPs(ctx context.Context, eps localEndpointCache, ciliumClient ciliumv2.CiliumV2Interface, enableCiliumEndpointSlice bool) error {
+	if enableCiliumEndpointSlice {
+		indexer := d.k8sWatcher.GetIndexer("ciliumendpointslice")
+		for _, cesObj := range indexer.List() {
+			ces, ok := cesObj.(*cilium_v2a1.CiliumEndpointSlice)
+			if !ok {
+				return fmt.Errorf("unexpected object type returned from ciliumendpointslice store: %T", cesObj)
+			}
+			for i := range ces.Endpoints {
+				cep := ces.Endpoints[i]
+				if cep.Networking.NodeIP == node.GetCiliumEndpointNodeIP() {
+					d.checkCiliumEndpoint(ctx, ces.Namespace, cep.Name, ciliumClient, eps)
+				}
+			}
+		}
+	} else {
+		store := d.k8sWatcher.GetStore("ciliumendpoint")
```

Comment:
Why don't we have an indexer for "ciliumendpoint"?

Reply:
This is a mistake, prior to refactoring I was listing Pods instead of CEPs directly which only watch for local Pods.

I'll add a similar indexer to the CES one.